api/src/pcapi/routes/backoffice_v3/forms/custom_reimbursement_rule.py

```diff
@@ -31,3 +38,55 @@ def is_empty(self) -> bool:
                 self.offerer.data,
             )
         )
+
+
+class EditCustomReimbursementRule(FlaskForm):
+    class Meta:
+        locales = ["fr_FR", "fr"]
+
+    offerer = fields.PCTomSelectField(
+        "Structure",
+        multiple=False,
+        choices=[],
+        validate_choice=False,
+        endpoint="backoffice_v3_web.autocomplete_offerers",
+    )
+    subcategories = fields.PCSelectMultipleField(
+        "Sous-catégories", choices=map(lambda s: (s.id, s.pro_label), subcategories_v2.ALL_SUBCATEGORIES)
+    )
+
+    rate = fields.PCDecimalField(
+        "Taux de remboursement %",
+        use_locale=True,
+        validators=[
+            wtforms.validators.InputRequired("Information obligatoire"),
+            wtforms.validators.NumberRange(min=0, message="Doit contenir un nombre positif"),
+            wtforms.validators.NumberRange(max=100, message="Doit contenir un nombre inférieur ou égal à 100"),
+        ],
+    )
+    start_date = fields.PCDateField("Date de début d'application")
+    end_date = fields.PCDateField(
+        "Date de fin d'application (optionnelle)", validators=(wtforms.validators.Optional(),)
+    )
+
+    def validate_start_date(self, start_date: fields.PCDateField) -> fields.PCDateField:
+        end_date_data = None
+        if self._fields["end_date"].data:
+            end_date_data = date_utils.get_day_start(self._fields["end_date"].data, finance_utils.ACCOUNTING_TIMEZONE)
+        start_date_data = date_utils.get_day_start(start_date.data, finance_utils.ACCOUNTING_TIMEZONE)
+
+        if end_date_data and (start_date_data > end_date_data):
+            raise wtforms.ValidationError("Ne peut pas être postérieure à la date de fin")
+
+        if start_date_data < date_utils.get_day_start(datetime.utcnow(), finance_utils.ACCOUNTING_TIMEZONE) + timedelta(
+            days=1
+        ):
+            raise wtforms.ValidationError("Ne peut pas commencer avant demain")
+
+        start_date.data = start_date_data
+        return start_date
+
+    def validate_end_date(self, end_date: fields.PCDateField) -> fields.PCDateField:
+        if end_date.data:
+            end_date.data = date_utils.get_day_start(end_date.data, finance_utils.ACCOUNTING_TIMEZONE)
```

Comment:
Attention, du coup tu reviens à 0:00, donc tu exclus le jour de fin, ce qui n'est probablement pas souhaité.
Par symétrie, une fonction `get_day_end` qui combine avec `datetime.max.time()` ?

Reply:
la pour le coup j'ai repris littéralement ce qui est dans flaskadmin

Reply:
En effet, du coup demandons à l'auteur, @dbaty était-ce voulu que la date de fin soit exclue, en fixant la fin à 0:00 cette date de fin, et non à 23:59 ?

Reply:
Hmm, on dirait effectivement qu'il y avait un bug. Bien vu ! Si l'utilisateur indique que le tarif est valable jusqu'au 31/01, il veut dire "le 31/01 à 23h59". Or, là, effectivement, on enregistre "31/01 à **0h**". Oups ! Heureusement, la plupart des tarifs (voire tous ?) ont été créés sans borne supérieure.

La borne supérieure de `CustomReimbursementRule.timespan` étant **ex**clusive, on pourrait écrire : 

```diff
- end_date.data = date_utils.get_day_start(end_date.data, finance_utils.ACCOUNTING_TIMEZONE)
+ end_date.data = date_utils.get_day_start(end_date.data + datetime.timedelta(days=1), finance_utils.ACCOUNTING_TIMEZONE)
```

Reply:
@dbaty J'ai peur que ta proposition change la date pour le lendemain dans le formulaire, et à l'affichage, ce qui créera de la confusion pour l'utilisateur

Reply:
Il faudrait alors faire cette conversion hors du validateur et ne pas modifier les données du formulaire.

Reply:
Attention, je ne parle pas uniquement de l'affichage dans le formulaire, mais aussi de l'affichage dans la liste des règles dans le backoffice.

Si on saisit un règle valide du 01/06/2023 au 30/06/2023 (dans le formulaire), alors avec `+ datetime.timedelta(days=1)` pour arriver 0h le lendemain (30 juin minuit), on stockera donc la plage du 01/06/2023 à 0:00:00 au 01/07/2023 à 0:00:00
et donc dans la liste des règles, avec ce formatage : 
https://github.com/pass-culture/pass-culture-main/blob/v238.0.1/api/src/pcapi/admin/custom_views/custom_reimbursement_rule_view.py#L155
on affichera comme plage de validité "01/06/2023 → 01/07/2023", l'utilisateur du backoffice sera surpris de voir que la date qu'il a saisie n'est pas respectée. De plus si on peut éditer la règle, on avancera d'un jour à chaque fois.
Je continue donc à militer pour s'arrêter le 30/06/2023 à 23:59:59 (donc `datetime.max.time()`)

Reply:
Oui, c'est juste. L'affichage actuel est incorrect (ça n'est pas dramatique, aucune règle active n'a de borne supérieure définie, actuellement). Deux possibilités : 

- soit on fait comme tu le proposes, mais il faudrait alors revoir la borne pour qu'elle ne soit plus **ex**clusive mais **in**clusive (même s'il est improbable qu'une réservation soit validée à 23:59:59.999999 !). Il faudrait alors peut-être aussi modifier 2-3 bouts de code en conséquence : au moins `CustomReimbursementRule.is_active()` et peut-être d'autres endroits.
- soit on laisse la borne exclusive et on adapte l'affichage (dans la page de détail et dans le champ rempli du formulaire d'édition).

Je n'ai pas d'avis très tranché sur la question. :)

J'aime bien quand la borne supérieure est exclusive (comme actuellement). Je trouve que ça facilite la lecture quand 2 règles se suivent : la borne supérieure de la première règle est égale à la borne inférieure de la seconde.

Reply:
En l'absence de @Meewan , je viens de faire les modifications dans cette PR : 
- dans les validateurs wtforms, on ne transforme plus en `datetime`, on garde simplement les dates, inclusives
- dans la route `create_custom_reimbursement_rules`, avant la création de l'objet `CustomReimbursementRule`, on fait la conversion en `datetime` qui était faite dans le validateur, avec une borne supérieure exclusive.
- dans `format_timespan`, pour l'affichage de la borne haute dans la liste, on retire une microseconde pour tomber en borne inclusive (et donc ne pas afficher le lendemain)
@dbaty si c'est bon pour toi par rapport aux changements demandés, peux-tu approuver avant la MES ?